source/dol/mLib/m_2d.cpp

```diff
@@ -0,0 +1,456 @@
+        }
+    } else {
+        frame += mRate;
+        if (mFlags & 1) {
+            if (frame > getLastActiveFrame()) {
+                frame = getLastActiveFrame();
+            }
+        } else if (frame >= mEndFrame) {
+            frame -= mEndFrame;
+        }
+    }
+    mCurrFrame = frame;
+}
+
+void m2d::FrameCtrl_c::set(float endFrame, u8 flags, float rate, float currFrame) {
+    if (currFrame < 0.0f) {
+        currFrame = 0.0f;
+    }
+    mEndFrame = endFrame;
+    mCurrFrame = currFrame;
+    setRate(rate);
+    mFlags = flags;
+    mPrevFrame = mCurrFrame;
+}
+
+void m2d::FrameCtrl_c::setFrame(float frame) {
+    mCurrFrame = frame;
+    mPrevFrame = frame;
+}
+
+void m2d::FrameCtrl_c::setRate(float rate) {
+    mRate = rate;
+}
+
+bool m2d::FrameCtrl_c::isStop() const {
+    switch (mFlags) {
+        case 1:
+            return mCurrFrame >= getLastActiveFrame();
+        case 3:
+            return mCurrFrame <= 0.0f;
+        default:
+            return false;
+    }
+}
+
+void m2d::Base_c::entry() {
+    m2d::Base_c *next = (m2d::Base_c *) nw4r::ut::List_GetNext(&m2d::l_list, nullptr);
+    while (next != nullptr) {
+        if (next->mDrawOrder > mDrawOrder) {
+            nw4r::ut::List_Insert(&m2d::l_list, next, this);
+            return;
+        }
+        next = (m2d::Base_c *) nw4r::ut::List_GetNext(&m2d::l_list, next);
+    }
+    nw4r::ut::List_Append(&m2d::l_list, this);
+}
+
+m2d::Simple_c::Simple_c() {
+    mpResAcc = nullptr;
+    mPos.set(0.0f, 0.0f, 0.0f);
+    mFlags = 0;
+    if (SCGetAspectRatio() == SC_ASPECT_WIDE) {
+        mDrawInfo.SetLocationAdjustScale(nw4r::math::VEC2(19.0f / 26.0f, 1.0f));
+        mDrawInfo.SetLocationAdjust(true);
+    }
+}
+
+void m2d::Simple_c::calc() {
+    calcBefore();
+    mMtx_c mtx;
+    PSMTXIdentity(mtx);
+    mVec3_c pos = mPos;
+    PSMTXTransApply(mtx, mtx, pos.x, pos.y, pos.z);
+    mDrawInfo.SetViewMtx(mtx);
+    calcAfter();
+}
+
+void m2d::Simple_c::calcBefore() {
+    u32 option = 0;
+    if (mFlags & 1) {
+        option = 1;
+    }
+    mLayout.Animate(option);
+}
+
+void m2d::Simple_c::calcAfter() {
+    mDrawInfo.SetViewRect(mLayout.GetLayoutRect());
+    mLayout.CalculateMtx(mDrawInfo);
+}
+
+void m2d::Simple_c::draw() {
+    nw4r::ut::Rect rect = mLayout.GetLayoutRect();
+    float near = 0.0f;
+    float far = 500.0f;
+    EGG::Screen screen;
+    bool isWide = EGG::Screen::sTVMode == EGG::Screen::TV_MODE_16_9;
+    float w_16_9 = EGG::Screen::sTVModeInfo[EGG::Screen::TV_MODE_16_9].width;
+    float w_4_3 = EGG::Screen::sTVModeInfo[EGG::Screen::TV_MODE_4_3].width;
+    float left = isWide ? w_16_9 * rect.left / w_4_3 : rect.left;
+    float right = isWide ? w_16_9 * rect.right / w_4_3 : rect.right;
+    screen.mProjType = EGG::Frustum::PROJ_ORTHO;
+    screen.ResetOrthographic(rect.top, rect.bottom, left, right, near, far);
+    if (isWide) {
+        screen.mScale = nw4r::math::VEC3(w_4_3 / w_16_9, 1.0f, 1.0f);
```

Comment:
```suggestion
    float width16x9 = EGG::Screen::sTVModeInfo[EGG::Screen::TV_MODE_16_9].width;
    float width4x3 = EGG::Screen::sTVModeInfo[EGG::Screen::TV_MODE_4_3].width;
    float left = isWide ? width16x9 * rect.left / width4x3 : rect.left;
    float right = isWide ? width16x9 * rect.right / width4x3 : rect.right;
    screen.mProjType = EGG::Frustum::PROJ_ORTHO;
    screen.ResetOrthographic(rect.top, rect.bottom, left, right, near, far);
    if (isWide) {
        screen.mScale = nw4r::math::VEC3(width4x3 / width16x9, 1.0f, 1.0f);
```
For readability (feel free to reject if you don't like it).